# Thông tin
- Họ tên: Nguyễn Phúc Thuần
- MSSV: 20120380

# Bài làm

In [154]:
def inverse_modulo(n, p):
    return pow(n, -1, p)

## 6.1
### (a)
Khóa công khai h của Alice là $h \equiv f^{-1}g \mod q \equiv 19928^{-1}18643 \equiv 767748560 \mod 918293817$

In [155]:
f = 19928
g = 18643
q = 918293817
h = (inverse_modulo(f, q) * g) % q
print(f'h = {h}')

h = 767748560


### (b)
Để giải mã, đầu tiên Alice cần tính 

$a \equiv fe \equiv 19928 \cdot 619168806 \equiv 600240756 \mod 918293817$

In [156]:
e = 619168806
a = f * e % q
print(f'a = {a}')

a = 600240756


Sau đó tính:

$m = f^{-1}a \equiv 19928^{-1}600240756 \equiv 11818 \mod g$

In [157]:
m = inverse_modulo(f, g) * a % g
print(f'm = {m}')

m = 11818


Plaintext cần tìm là $m = 11818$

### (c)
Ta có $e \equiv rh + m \mod q \equiv 19564\cdot767748560 + 10220 \mod 918293817 \equiv 619167208 \mod 619167208$

In [158]:
r = 19564
m = 10220
e = (r*h + m) % q
print(f'e = {e}')

e = 619167208


Ciphertext Bob gửi cho Alice là $e = 619167208$

## 6.3


Ta biết rằng $M_i = Ar_i\mod B$ với $r_n < \frac{B}{2}$. Do đó Eve có thể suy ra $r_i \equiv \dfrac{M_i}{A}\mod{B}$

In [159]:
def resolve_binary(M, A, B):
    r = []
    for sack in M:
        r.append(sack * inverse_modulo(A, B) % B)
        
    return r

In [160]:
M = (5186, 2779, 5955, 2307, 6599, 6771, 6296, 7306, 4115, 7039)
A = 4392
B = 8387
S = 26560
r = resolve_binary(M, A, B)
print(f'r = {r}')

r = [5, 14, 30, 75, 160, 351, 750, 1579, 3253, 6500]


Bằng cách trên Eve tìm được superincreasing sequence r = {5, 14, 30, 75, 160, 351, 750, 1579, 3253, 6500}

Eve tiến hành tính $S' = A^{-1}S \mod{B} \equiv 4528\mod{8387}$

In [161]:
S1 = inverse_modulo(A, B) * S % B
print(f'S\' = {S1}')

S' = 4528


Cuối cùng, với thuật toán đã nêu ở phần Proposition 6.5, Eve tính x:

In [162]:
def solve_subset(M, S):
    n = len(M)
    x = [0] * n
    for i in reversed(range(n)):
        if S >= M[i]:
            x[i] = 1
            S = S - r[i]
            
    return x

In [163]:
x = solve_subset(r, S1)
print(f'x = {x}')

x = [0, 1, 0, 0, 1, 1, 1, 0, 1, 0]


Vậy plaintext x mà Bob gửi cho Alice là x = {0, 1, 0, 0, 1, 1, 1, 0, 1, 0}

## 6.6

In [164]:
import numpy as np

def get_orthogonal(V):

    n = V.shape[1]
    for j in range(n):

        for k in range(j):
            V[:, j] -= np.dot(V[:, k], V[:, j]) * V[:, k]
        V[:, j] = V[:, j] / np.linalg.norm(V[:, j])
    return V

### (a)

In [165]:
V = np.array([[1.0, 3.0, 2.0], [4.0, 1.0, -2.0], [-2.0, 1.0, 3.0]])

V_star = get_orthogonal(V)
print(f'V_star = {V_star}')

V_star = [[ 0.21821789  0.88182973 -0.41803981]
 [ 0.87287156  0.01520396  0.48771311]
 [-0.43643578  0.47132279  0.76640632]]


### (b)

In [166]:
V = np.array([[4, 1, 3, -1], [2, 1, -3, 4], [1.0, 0, -2, 7]])

V_star = get_orthogonal(V)
print(f'V_star = {V_star}')

V_star = [[ 0.87287156 -0.26726124  0.40824829 -0.57735027]
 [ 0.43643578  0.80178373 -0.40824829 -0.57735027]
 [ 0.21821789 -0.53452248 -0.81649658 -0.57735027]]


## 6.10
### b, det(A) ≠ 0. Chứng minh $A^{-1} = \dfrac{1}{det(A)}B$

Xét tính chất của phần tử kề của ma trận vuông

Ta có $det(A) = A.B$ 

$\Rightarrow   A^{-1}det(A) = A^{-1}AB \\
\Leftrightarrow  A^{-1}det(A) = I_nB \\
\Leftrightarrow  A^{-1} = \dfrac{B}{det(A)}$ (đpcm)

### a, Chứng minh $AB = BA = det(A)I_n$
Ta mượn kết quả đã chứng minh ở câu b:

$
A^{-1} = B/det(A)
\Leftrightarrow  A.A^{-1} = (A.B)/det(A)
\Leftrightarrow I = (A.B)/det(A)$ (1)

và 

$A^{-1} . A = \dfrac{BA}{det(A)}
\Leftrightarrow  I = \dfrac{BA}{det(A)}$ (2)

Từ (1) và (2) suy ra: $I_n = \dfrac{AB}{det(A)} = \dfrac{BA}{det(A)}$

Hay $AB = BA = det(A)I_n$ (đpcm)
